### ISSUE TITLE
Mysterious case error in DirAsDsk (WINDOWS)

### ISSUE THREAD

**Issue by uniskie** [openMSX/openMSX#1568]
This is a problem with the Windows version of OpenMSX.

When using DirAsDsk, I save the file in lower case, but if there is a file name with upper case letters, an error continues to occur.
(If you remount it with DirAsDsk, the error will disappear.)

1. I wonder why it is saved in lower case. Files brought from Real MSX are in uppercase.

2. Why does OpenMSX keep giving error messages even though there are no problems with the Windows file system?

![image](https://github.com/openMSX/openMSX/assets/11275674/29d7d361-3dd1-441f-87df-0e5e1d86ab60)

**Comment by MBilderbeek** [comment]
Hi,

Thanks for reporting.

We use lower case, because it looks much better on the host system. (Who likes SCREAMING NAMES?)

Can you be a bit more clear on the scenario on how to reproduce this? So which steps (from a blank directory onwards) are you taking to trigger this situation?

**Comment by uniskie** [comment]
Thanks for the reply.
I am always grateful for your help.

Due to a problem with the translation software, the explanation did not seem to come across well.

Operations on empty directories are not a problem.
The problem occurs when overwriting a file copied from an actual MSX disk. (Only when using DirAsDsk)

I will explain it step by step.

## Steps:

1. Copy the files in the 2DD Floppy DIsk for MSX (or contents of dsk file) to the directory on Windows.
2. Make sure that the file names of the files you copied from the MSX disk are in uppercase.
3. Mount the copied directory on Windows using DirAsDsk in OpenMSX. (Use Catapult's Disk Drives or console's dska or dskb)  
4.  Overwrites a file that already exists on DiskDrive with the same name from OpenMSX emulation.
5. Every time I perform a disk operation, I get the error "Couldm't add (filename): (filename) already exists".
6. Also confirm that the overwriting of the file was actually successful.
   (It will be easier to understand if you change the content to be saved.)

## Additional notes:

- Remounting the directory will eliminate the error.
- No error occurs if the file name on host is lowercase.
- Files visible from MSX may multiply into two.
  1. the file name on the host is lower case
  2. mounted with DirAsDsk
  3. rewrite the file name to upper case from the host PC during mounting
  4. execute files or dir command on msx. 
  5. Sometimes it appears as if there are two files from MSX emulation, even though there is only one file on the host.
- I think these problems occurs in OpenMSX for Windows (MS-DOS series).


## About upper and lowercase letters:

This is just my opinion, but if you implement Directory as a DSK as an extension of MSX emulation, I think it would be more natural to capitalize it to match MSX. 
This is because if you copy a file from 2DD for MSX, it will be in uppercase.

I think there is a difference of opinion between working with actual equipment and not working with it.

However, the more important issue is that I keep getting inappropriate error messages.

**Comment by MBilderbeek** [comment]
I tried this, please let me know whether it's the correct scenario:
1. I create a folder test
2. I copy a file with uppercase letters in it, e.g. `KINK.BAS`
3. I start openMSX with this folder and confirm it shows up in MSX-BASIC when I execute the `FILES` command
4. I type in MSX-BASIC: `save kink.bas` (to save an empty program on top of it).
5. I execute `FILES` again, but I still only see `KINK.BAS`, which is now empty. On the host OS, I also see one file, upper case `KINK.BAS`

Is this the correct scenario, or did I misunderstand? If it's the correct scenario, you see that I cannot reproduce the problem, alas.

I also did an extra test:
1. on the host, I renamed `KINK.BAS` to `kink.bas`
2. I executed `FILES` again in MSX-BASIC and it still shows only one time `KINK.BAS` and on the host it also still shows `kink.bas`.
So also this problem I cannot reproduce.

I am not using Windows, but Linux. Maybe it's different. Also, I'm using the latest development build.

**Comment by uniskie** [comment]
Thank you for testing.

The version and operating procedure are correct,
This probably does not occur on Linux.

This phenomenon is considered to be Windows-specific.

(I suppose that if you had multiple files on Linux that differed only in case, MSX would have a problem distinguishing between them, but that would be a different issue. But that's a different issue.)

It does not occur when using DSK images, only when using DirAsDsk.

This is an imaginary scenario, but it could be that the case-sensitivity in the process of comparing the internal file list with the filenames of the host files is causing the error to be displayed.

I am not sure if this is relevant,
There is also a problem with the language locale of the HOST PC not being converted properly when using DirAsDsk. (In Japanese, multi-byte characters are garbled.)

DirAsDsk is a very useful feature that I use all the time. It's a great feature that I can't do without.

**Comment by fhil** [comment]
I've done some digging on this problem.

![image](https://github.com/user-attachments/assets/ad089574-507c-45aa-ae04-e86e1acc9d0b)

![image](https://github.com/user-attachments/assets/fd063f7b-574d-48a8-9d7c-79b5281eb146)

![image](https://github.com/user-attachments/assets/955989fd-74b9-40ee-9bd1-818be18f623f)

![image](https://github.com/user-attachments/assets/84a4f848-450b-4605-a8d7-51da77d5b0ed)

**Comment by fhil** [comment]
my local directory contains some files:
- A.BAT
- B.BAT
- I.BAT
- GT.PAS
The local files are added as directory entries to sector 5.
A.BAT will be on sector 5 and index 0.
B.BAT will be on sector 5 and index 1.
I.BAT will on on sector 5 and index 2.
etc....

So i let the msx boot up (configuration is a vg8235) and do some command on dos to write to the file B.BAT.
![image](https://github.com/user-attachments/assets/8e9fc2d4-6008-4a57-8a13-77cc5fbf0fb2)

The function scanMsxDirs() finds the correct result which is sector 5 and index 1.
Just after leaving that function the result is gone. Sector and Index both become 0.

Which causes sector 0 index 0 to be interpreted as a directory entry, which it isn't. And that causes openmsx to crash almost immediately.

Why isn't scanMsxDirs() returning its value?

![image](https://github.com/user-attachments/assets/639eff6b-cf54-4bd4-aef8-62dfabc2312f)

I've added a watch on p_scanner to see it there would be any change, but there isn't any change made to scanner
I think that the call to scanMsxDirs(scanner, firstDirSector)) uses the variable scanner as a reference instead of a value.

Anyway. Maybe someone can comment on this?

**Comment by fhil** [comment]
I'm using visual studio 2022 on a windows 11 system

**Comment by uniskie** [comment]
(to fhil)
I cannot answer your question on that issue.

The topics in this thread are,

1) OpenMSX warns me even though Windows Host is case-insensitive for filenames.
     → I would like for Windows to be case-insensitive in filenames and not issue warnings.

2) When overwriting a file, I want it to match the case of the original filename.
    (Since MSX is also case-insensitive for file names)

This was the contents of the request.

**Comment by fhil** [comment]
Yes that was the original request. but i had a version from around 19th of may this year and that version crashed the way i described while testing what was going on.
Wouter seems to have fixed this call by value problem on 31th of may. I didn't check for that update.

Can you try it again with a recent build?

**Comment by uniskie** [comment]
I have built and tested the latest version and the file name warning issue remains the same.

I am using MSX-DOS1 so I don't know if it is cured about the crash on the directory name check.

**Comment by fhil** [comment]
![image](https://github.com/user-attachments/assets/912f1afe-5ea6-439a-a8d6-fe1b3d4e1058)

I booted into basic and create an empty file with:
   save"test.bas"
Then i renamed test.bas to TEST.BAS from within windows
Then did another:
   save"test.bas"

this gives the error messages in the picture

**Comment by fhil** [comment]
i think that this gives the same problem on Linux ?

**Comment by m9710797** [comment]
fhil, uniskie,

Thanks for your comments. I've been busy, I didn't have time yet to look into this. I hope to have more time next weekend.

**Comment by MBilderbeek** [comment]
@fhil I tried this on Linux with the latest openMSX version, but there was no error message at all and the file was still called TEST.BAS (and the content has actually changed with what I saved it to).
